icefoganalytics/travel-authorization
# Ytgov Issue 90: Travel Auth Reassign a Supervisor/Approver

Fixes https://github.com/ytgov/travel-authorization/issues/90

Relates to:

- Travel Auth approval
- signing authority validation

# Context

**Is your feature request related to a problem? Please describe.** An assigned approver may not have the proper signing authority to approve the travel authorization form. The assigned supervisor may need to reassign it to the proper authority to finalize the approval.

**Describe the solution you'd like** Ideally the traveler will be told who to assign it to. The traveler may choose to assign it to their supervisor in advance of the signing authority. The supervisor can approve the travel with comments and assign it either to the signing authority or on to another supervisor. The final approval must come from the coding authority (we may need to capture the coding in this screen to define who is the authority).

**Describe alternatives you've considered** Rather than integrating the signing authority api at this point. Allow supervisors to reassign the ta to another person. Add verbiage to the approve button that says something like "by selecting approval it has been acknowledged that you have recieved the proper approval from the proper signing authority " (better words to come if need to use this solution)

**Additional context** I am not sure the status of the signing auth api. May need to confirm we can use this.

# Implementation

1. Support re-assignment of travel authorization to another user via "submit" endpoint.
2. Submit endpoint now accepts nested  travelAuthorizationActionLogAttributes -> note.
3. Enhance snack plugin so that it supports snack.<variant>(message, options) syntax. This makes it easier to write a snack.success(message) or snack.error(message).

# Screenshots

Travel Authorization management re-assignment buttons and log
http://localhost:8080/travel-requests/9/details/manage
![image](https://github.com/user-attachments/assets/625e719e-9c37-45ac-93a2-719f12dc1e65)

Travel Authorization management re-assignment dialog
http://localhost:8080/travel-requests/9/details/manage?showReAssign=true
![image](https://github.com/user-attachments/assets/4f6f454e-78ed-4f32-91a7-44fa461055ae)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Requests" page via the to dropdown nav.
5. Click on the "+ Travel Authorization" button to create a travel authorization.
6. Fill out the form and generate an estimate.
7. Submit the travel authorization to a supervisor.
8. Log in as the supervisor.
9. Go to the "Manager View" page via the top dropdown nav.
10. Click on the relevant travel authorization in the "Pending Approvals" card.
11. Scroll to the bottom of the page and click the "Re-assign" button to open the re-assigment dialog.
12. Select a user to re-assign the travel authorization to and add a note.
13. Click the "Re-assign" button to re-assign the travel authorization and be redirected back to the "Manager View" page.
14. Log in as the user that the travel authorization was re-assigned to.
15. Go to the "Manager View" page via the top dropdown nav.
16. Click on the relevant travel authorization in the "Pending Approvals" card.
17. Scroll to the bottom of the page and see the new log entry with the re-assignment note.
